src/thor/chinese_postman_action.cc

```diff
@@ -0,0 +1,484 @@
+      return i;
+    }
+    i++;
+  }
+  return i;
+}
+
+std::vector<PathInfo> buildPath(GraphReader& graphreader,
+                                const Options& /*options*/,
+                                const valhalla::Location& origin,
+                                const valhalla::Location& dest,
+                                const baldr::TimeInfo& time_info,
+                                const bool invariant,
+                                std::vector<GraphId> path_edges,
+                                const std::shared_ptr<sif::DynamicCost>& costing_,
+                                float source_pct,
+                                float target_pct) {
+  // Build a vector of path info
+  // once we recovered the whole path we should construct list of PathInfo objects
+  // set of edges recovered from shortcuts (excluding shortcut's start edges)
+  std::unordered_set<GraphId> recovered_inner_edges;
+
+  std::vector<PathInfo> path;
+  path.reserve(path_edges.size());
+
+  auto edge_itr = path_edges.begin();
+  const auto edge_cb = [&edge_itr, &path_edges]() {
+    return (edge_itr == path_edges.end()) ? GraphId{} : (*edge_itr++);
+  };
+
+  const auto label_cb = [&path, &recovered_inner_edges](const EdgeLabel& label) {
+    path.emplace_back(label.mode(), label.cost(), label.edgeid(), 0, label.path_distance(),
+                      label.restriction_idx(), label.transition_cost(),
+                      recovered_inner_edges.count(label.edgeid()));
+  };
+
+  // recost edges in final path; ignore access restrictions
+  try {
+    sif::recost_forward(graphreader, *costing_, edge_cb, label_cb, source_pct, target_pct, time_info,
+                        invariant, true);
+  } catch (const std::exception& e) {
+    LOG_ERROR(std::string("Chinese Postman failed to recost final path: ") + e.what());
+  }
+
+  return path;
+}
+
+DistanceMatrix thor_worker_t::computeCostMatrix(std::vector<baldr::GraphId> graph_ids,
+                                                const std::shared_ptr<sif::DynamicCost>& costing,
+                                                const float max_matrix_distance) {
+  DistanceMatrix distanceMatrix(boost::extents[graph_ids.size()][graph_ids.size()]);
+  if (graph_ids.size() == 0) {
+    return distanceMatrix;
+  }
+
+  // TODO: Need to populate these two variables
+  google::protobuf::RepeatedPtrField<valhalla::Location> source_location_list;
+  google::protobuf::RepeatedPtrField<valhalla::Location> target_location_list;
+
+  for (const auto& graph_id : graph_ids) {
+    GraphId g(graph_id);
+    auto l = getPointLL(g, *reader);
+    Location loc = Location();
+
+    loc.mutable_ll()->set_lng(l.first);
+    loc.mutable_ll()->set_lat(l.second);
+    source_location_list.Add()->CopyFrom(loc);
+    target_location_list.Add()->CopyFrom(loc);
+  }
+
+  try {
+    auto locations = PathLocation::fromPBF(source_location_list, true);
+    const auto projections = loki::Search(locations, *reader, costing);
```

Comment:
hm not sure it's good to call loki from thor..

Reply:
Any suggestion for this one @nilsnolde @kevinkreiser ? I am out of idea for this one

Reply:
yeah this will be straight forward. ill stub in a method for you later tonight, perhaps i can even make it so that loki::Search could reuse the method that i stub it when it detects a "node-snapped" projection, then we can get a little bit of code re-use out of it too!